File: leetcode/101-200/123-best-time-to-buy-and-sell-stock-iii/bench/max_profit.c
/* Benchmark workload — at-most-two-transactions DP, Best Time to Buy and Sell Stock III.
 * Algorithmic mirror of bench/max_profit.kara. See ../README.md § Benchmarks for N / K and the LCG. */
#include <stdio.h>
#include <stdlib.h>

static long long max_profit(const long long *prices, long long n) {
    if (n == 0) return 0;
    long long buy1 = -prices[0], sell1 = 0, buy2 = -prices[0], sell2 = 0;
    for (long long i = 1; i < n; i++) {
        long long p = prices[i];
        if (-p > buy1) buy1 = -p;
        if (buy1 + p > sell1) sell1 = buy1 + p;
        if (sell1 - p > buy2) buy2 = sell1 - p;
        if (buy2 + p > sell2) sell2 = buy2 + p;
    }
    return sell2;
}

int main(void) {
    long long n = 2000000;
    long long *data = (long long *)malloc(sizeof(long long) * n);
    long long state = 12345;
    for (long long i = 0; i < n; i++) { state = (state * 1103515245 + 12345) & 2147483647; data[i] = (state & 4095) + 1; }
    long long sum = 0;
    for (int k = 0; k < 10; k++) {
        long long r = max_profit(data, n);
        sum += r;
        data[0] = ((data[0] + r) & 4095) + 1;
    }
    printf("%lld\n", sum);
    free(data);
    return 0;
}

File: leetcode/101-200/123-best-time-to-buy-and-sell-stock-iii/bench/max_profit.rs
//! Benchmark workload — at-most-two-transactions DP, Best Time to Buy and Sell Stock III.
//!
//! Algorithmic mirror of bench/max_profit.kara. See ../README.md § Benchmarks for N / K and the LCG.
//! Serial K-loop (each rep perturbs the opening price by the previous result) so all mirrors run the
//! same single-threaded dependency chain. Kara checks integer overflow by default, so the
//! like-for-like row is `rustc -O -C overflow-checks=on`.

fn max_profit(prices: &[i64]) -> i64 {
    if prices.is_empty() {
        return 0;
    }
    let (mut buy1, mut sell1, mut buy2, mut sell2) = (-prices[0], 0i64, -prices[0], 0i64);
    for i in 1..prices.len() {
        let p = prices[i];
        if -p > buy1 {
            buy1 = -p;
        }
        if buy1 + p > sell1 {
            sell1 = buy1 + p;
        }
        if sell1 - p > buy2 {
            buy2 = sell1 - p;
        }
        if buy2 + p > sell2 {
            sell2 = buy2 + p;
        }
    }
    sell2
}

fn main() {
    const N: usize = 2_000_000;
    let mut data: Vec<i64> = Vec::with_capacity(N);
    let mut state: i64 = 12345;
    for _ in 0..N {
        state = (state.wrapping_mul(1103515245).wrapping_add(12345)) & 2_147_483_647;
        data.push((state & 4095) + 1);
    }
    let mut sum: i64 = 0;
    for _ in 0..10 {
        let r = max_profit(&data);
        sum += r;
        data[0] = ((data[0] + r) & 4095) + 1;
    }
    println!("{}", sum);
}
